### PR TITLE
feat: add dependency graph

### LINKED ISSUE #11: feat: dependency graph
current implementation does find all components.
current implementation does not populate the dependency graph

## feature request

have the dependency graph populated.

### PR DESCRIPTION
**Pull request by Churro**
### Description

Implement the CycloneDX dependency graph by analyzing esbuild/Bun metafile import relationships.

The algorithm walks `metafile.inputs`, maps each file to its owning package component, then follows imports edges to build package-level dependency relationships. Only packages with actual bytes in the final bundle output are included as components. Tree-shaken packages are excluded from the SBOM.

Three import path resolution strategies handle bundler differences:

1. **Direct file-path lookup:** works for esbuild's relative paths
2. **Absolute-to-relative normalization:** handles Bun's absolute import paths
3. `createRequire().resolve()` **fallback:** handles Bun's unresolved bare specifiers (e.g. `react`)

The components lookup map is expanded to cover tree-shaken files within bundled packages, so import edges through zero-byte entry points (e.g. `react/index.js` re-exporting from `react/cjs/react.development.js`) are not missed.

Virtual files (e.g. `angular:polyfills:...`) are naturally filtered out since they have no `package.json` and never become components.

Resolves or fixes issue: #11 

### AI Tool Disclosure

- [ ] My contribution does not include any AI-generated content
- [x] My contribution includes AI-generated content, as disclosed below:
  - AI Tools: `OpenCode`
  - LLMs and versions: `Claude Sonnet 4.6`
  - Prompts: `Implement the dependency graph feature for cyclonedx-esbuild (issue #11). Research esbuild and Bun metafile structures from source code. Design and implement the linkDependencies algorithm with correctness proofs. Handle Bun compatibility (absolute paths, bare specifiers). Verify against all testbeds (angular20, juice-shop, react19-bun, typescript-npm, typescript-yarn, esbuild-latest).`

### Affirmation

- [x] My code follows the [CONTRIBUTING.md](https://github.com/CycloneDX//cyclonedx-esbuild/blob/main/CONTRIBUTING.md) guidelines

### CODE REVIEW

**Comment by jkowalleck** [comment]
Thanks for championing this topic, @Churro . :+1: 

I might need some time to review, so please keep your pull-request open and don't delete/archive your fork of the repository.

File: src/builders.ts [hunk 1]
```diff
@@ -155,19 +156,59 @@ export class BomBuilder {
...
+    // Expand the components map to cover all metafile.inputs files that belong to known packages.
+    // This is needed because import targets may point to tree-shaken files (bytesInOutput=0)
+    // within packages that do have other files in the bundle.
+    for (const filePath of Object.keys(metafile.inputs)) {
+      if (components.has(filePath)) continue
+      const pkg = getPackageConfig(resolve(rootDir, filePath))
+      if (pkg === undefined) continue
+      const component = pkgs.get(pkg.path)
+      if (component === undefined) continue
+      components.set(filePath, component)
```
Comment (jkowalleck) [inline src/builders.ts]: is this about adding components that are not bundled due to tree-shaking? 
If yes, then the foollowing apllies:

---

my first idea was: these components shall be marked as "excluded" since they will be not in the build result due to tree shaking.
see https://cyclonedx.org/docs/1.7/json/#metadata_tools_oneOf_i0_components_items_scope

when we set the scope for these here, then we should set also set the scope of the actually bundled components to `Required` first and fore-most.

```suggestion
      components.scope ??= ComponentScope.Excluded
      components.set(filePath, component)
```

----

need to look onto the actual details, when doing a proper review.

File: src/builders.ts [hunk 2]
```diff
@@ -155,19 +156,59 @@ export class BomBuilder {
...
+    }
+    this.linkDependencies(metafile, components, pkgs, rootDir)
 
     logger.info(LogPrefixes.INFO, 'done building Components from modules...')
     return pkgs
   }
 
-  /* @ts-expect-error TS6133 -- TODO */
-  private linkDependencies(metafile: esbuild.Metafile, modulesComponents: Map<string, Component>): void {
-    // TODO: link deps based on inputs - https://github.com/CycloneDX/cyclonedx-esbuild/issues/11
```
Comment (jkowalleck) [inline src/builders.ts:224]: note for proper review: see how this plan/idea worked out

File: src/builders.ts [hunk 3]
```diff
@@ -155,19 +156,59 @@ export class BomBuilder {
     }
 
     logger.info(LogPrefixes.INFO, `linking Component.dependencies...`)
-    this.linkDependencies(metafile, components)
+    // Expand the components map to cover all metafile.inputs files that belong to known packages.
+    // This is needed because import targets may point to tree-shaken files (bytesInOutput=0)
+    // within packages that do have other files in the bundle.
+    for (const filePath of Object.keys(metafile.inputs)) {
+      if (components.has(filePath)) continue
```
Comment (jkowalleck) [inline src/builders.ts]: note for proper review: 
see if we can craft a test-bed/case where all these code branches are actually hit...

File: src/builders.ts (hunk 1 above)
Comment (Churro) [reply src/builders.ts]: No, we don't add three-shaken / non-bundled components components here. In that case (all files have 0 bytes), it would never appear in `pkgs` in the first place, so `pkgs.get(pkg.path)` returns undefined and we continue (see a few lines before). 

The purpose of this expansion is to map a tree-shaken file to an already-bundled component. In other words, we are making an existing, bundled component findable by additional file paths. Setting `components.scope ??= ComponentScope.Excluded` would be setting scope on an already-bundled component, which would be incorrect.

Perhaps an example helps:
1. Package react has files `index.js` (re-export, 0 bytes in output) and `cjs/react.development.js` (actual code, >0 bytes)
2. react _is_ in the SBOM because `cjs/react.development.js` has bytes and the component was already created
3. react-dom imports `react/index.js` (the 0-byte file)
4. Without the expansion that's done here, `components.get("node_modules/react/index.js")` returns undefined, so the react-dom -> react dependency edge is missed
5. To fix this, we add a mapping `node_modules/react/index.js` -> react component to the lookup map so the edge resolves

I found this while verifying the react19-bun testbed but the same pattern occurs basically in any package that uses a re-export `index.js`.

File: src/builders.ts [hunk 4]
```diff
@@ -155,19 +156,59 @@ export class BomBuilder {
...
   }
 
-  /* @ts-expect-error TS6133 -- TODO */
-  private linkDependencies(metafile: esbuild.Metafile, modulesComponents: Map<string, Component>): void {
-    // TODO: link deps based on inputs - https://github.com/CycloneDX/cyclonedx-esbuild/issues/11
-    // idea: take the metadata.input
-    // then cut the "externals" and copy their content to all the ones that used it
-    // then cut the "unknown" and copy their content to all the ones that used it
-    // the rest should all be known components -> so set their dependencies as expected
+  private linkDependencies(
```
Comment (Churro) [inline src/builders.ts:212]: I didn't want to over-comment but perhaps this helps with the review to describe the approach once more in natural language:

For each `filePath` in `metafile.inputs`, find its owning package component (`sourceComponent`), follow its imports to find target package components. Therefore, we use 3 import path resolution strategies to address differences among bundlers (as also noted in the initial PR description):

   1. Direct lookup in the file -> component map (works for esbuild's relative paths)
   2. Absolute-to-relative normalization (handles Bun's absolute import paths)
   3. `createRequire().resolve()` fallback (handles Bun's unresolved bare specifiers)

Self-edges (`source === target`) are filtered to avoid intra-package noise.
External imports are skipped as they are not part of the bundle.

**Comment by Churro** [comment]
@jkowalleck, do you see a chance we could continue with this anytime soon?

I understand that the change is not trivial to review but since this adds functionality that hasn't been there until now, I'd advocate to treat it as a step into the right direction with fixes or corrections being possible later on too, if needed 😊

**Comment by jkowalleck** [comment]
> @jkowalleck, do you see a chance we could continue with this anytime soon?

Thanks a lot for your patience and for pushing this forward.

I can’t promise anything concrete right now, but a detailed review is planned for this month. What I _can_ offer in the meantime is to publish a beta release containing the current state of the feature, so you can battle‑test it in real projects and help surface any edge cases early.

Please understand that SBOM generation and dependency‑tree accuracy are critical for many consumers of this package. Producing incomplete or misleading data can erode trust in both the tool and the wider ecosystem. That’s why changes in this area require a very thorough review — regardless of whether the contribution is AI‑assisted or fully human‑written.

Thanks again for your work and for staying engaged with the process.

File: src/builders.ts [hunk 5]
```diff
@@ -155,19 +156,59 @@ export class BomBuilder {
...
     logger.info(LogPrefixes.INFO, `linking Component.dependencies...`)
-    this.linkDependencies(metafile, components)
+    // Expand the components map to cover all metafile.inputs files that belong to known packages.
+    // This is needed because import targets may point to tree-shaken files (bytesInOutput=0)
+    // within packages that do have other files in the bundle.
+    for (const filePath of Object.keys(metafile.inputs)) {
+      if (components.has(filePath)) continue
+      const pkg = getPackageConfig(resolve(rootDir, filePath))
+      if (pkg === undefined) continue
+      const component = pkgs.get(pkg.path)
```
[the remark this answers is not in the document]
Comment (jkowalleck) [reply src/builders.ts]: adressed in #20

File: src/builders.ts [hunk 6]
```diff
@@ -155,19 +156,59 @@ export class BomBuilder {
...
+            const contextDir = dirname(resolve(rootDir, filePath))
+            const resolved = createRequire(resolve(contextDir, '_')).resolve(imp.path)
+            const pkg = getPackageConfig(resolved)
+            if (pkg !== undefined) {
+              targetComponent = pkgComponents.get(pkg.path)
+            }
+          } catch {
+            // Module not resolvable from this context — skip this edge
+          }
+        }
```
[the remark this answers is not in the document]
Comment (jkowalleck) [reply src/builders.ts]: addressed in #20

**Comment by Churro** [comment]
> I can’t promise anything concrete right now, but a detailed review is planned for this month. What I _can_ offer in the meantime is to publish a beta release containing the current state of the feature, so you can battle‑test it in real projects and help surface any edge cases early.

Thanks! I appreciate the idea of a beta release but if a detailed review is just a matter of a few weeks, I'm happy to wait.

Meanwhile, regarding https://github.com/CycloneDX/cyclonedx-esbuild/pull/113#discussion_r3141750650 I took up the suggestion of a testbed and added one in https://github.com/CycloneDX/cyclonedx-esbuild/pull/113/commits/d212cc0f5a6f27a0fc4498a2f54fc07b51099f61 that covers specific branches in `linkDependencies` and the expansion loop that no existing testbed hits:

* **External import** (`node:path` marked as `external`): hits `if (imp.external === true) return undefined` in `resolveImportComponent`. Without this, that branch is dead code in tests.
* **Entirely tree-shaken package (`unused-package`)**:  imported in source but never used, so esbuild tree-shakes all its code. It appears in metafile.inputs but has `bytesInOutput=0` everywhere and is never added to pkgs. This hits the if `(component === undefined) continue` branch in the expansion loop, the case where a reachable package is completely absent from the bundle.
* **Zero-byte entry point of a bundled package (`custom-package/lib/index.js`)**: `index.js` re-exports from `utils.js`, so esbuild inlines `utils.js` directly and leaves `index.js` with 0 bytes. But the import edge from `src/index.mjs` still points to `index.js`. This hits the expansion loop's happy path: adding a zero-byte file to the lookup map so the dependency edge app → custom-package isn't missed.
* **Intra-package self-edge filter**: `custom-package/lib/index.js` imports `utils.js` and `used_private.js`, all same package. Hits `if (targetComponent === sourceComponent) continue`.

So overall with this testbed, only the branch where `createRequire` throws remains uncovered. This is a Bun-specific defensive path where bare specifiers point to nonexistent modules. Not reproducible with esbuild without contrived plugin tricks.

File: src/builders.ts (hunk 1 above)
Comment (jkowalleck) [reply src/builders.ts]: i see. so the issue lies here: https://github.com/CycloneDX/cyclonedx-esbuild/blob/ce40c02ea816de91bff7a1decfe38f383cbfe1f1/src/builders.ts#L127
this would skip components that are tree-shaken.

I will create a ticket to keep tree-shaken components in the component list, but mark them as "excluded".
Will work on this asap.

Comment (jkowalleck) [reply src/builders.ts]: this discussion caused https://github.com/CycloneDX/cyclonedx-esbuild/issues/142

File: src/builders.ts [hunk 7]
```diff
@@ -191,19 +192,69 @@ export class BomBuilder {
     }
 
     logger.info(LogPrefixes.INFO, `linking Component.dependencies...`)
-    this.linkDependencies(metafile, components)
+    // Expand the components map to cover all metafile.inputs files that belong to known packages.
+    // This is needed because import targets may point to tree-shaken files (bytesInOutput=0)
+    // within packages that do have other files in the bundle.
+    for (const filePath of Object.keys(metafile.inputs)) {
```
Comment (jkowalleck) [inline src/builders.ts]: this can be reverted, since #142 is done.

**Review by jkowalleck — CHANGES_REQUESTED** [review]

File: src/builders.ts (hunk 4 above)
Comment (jkowalleck) [reply src/builders.ts:212]: need to craft test cases for external packages.

PS: done via #150 / #151

File: src/builders.ts (hunk 7 above)
Comment (jkowalleck) [reply src/builders.ts]: done via 2a4d167052b769ec1ee38b6799f79827cfe104c5

File: tests/_data/unused-package/package.json
Comment (jkowalleck) [inline tests/_data/unused-package/package.json:1]: might move this out of this very pull-request, so we have a proper diff/before-after-comparison

Comment (jkowalleck) [reply tests/_data/unused-package/package.json:1]: done via https://github.com/CycloneDX/cyclonedx-esbuild/pull/152

File: tests/_testbeds/dep-graph-coverage/package.json
Comment (jkowalleck) [inline tests/_testbeds/dep-graph-coverage/package.json:1]: will remove this package.
there actually are no edge cases, just cases. and they were added via #151 and #152

**Review by jkowalleck — CHANGES_REQUESTED** [review]

**Comment by jkowalleck** [comment]
all those extra tests were moved out of this PR into individual contributions. (#144, #151, #152).

and some parts of the implementation were removed as they were no longer needed since #143 was done.

need to review the rest of the implementation, still.



@Churro, next time, please dont try to solve all things at once.
Better have test-cases defined first and add them in individual contributions - and add new functionality on top of existing tests. 
This makes it a lot easier to see how a feature behaves and how it changes results. :-)

**Comment by jkowalleck** [comment]
continues in https://github.com/CycloneDX/cyclonedx-esbuild/pull/161

**Comment by jkowalleck** [comment]
was released via https://github.com/CycloneDX/cyclonedx-esbuild/releases/tag/v1.4.0
